File: 0_1.baekjoon/tree/java/src/BOJ17037.java
import java.io.*;
import java.util.*;

public class BOJ17037 {
    private static int N;
    private static double W;
    private static int[] degree;
    public static void main(String args[]) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        W = Double.parseDouble(st.nextToken());
        degree = new int[N + 1];

        for (int i=0;i<N-1;i++) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());

            degree[u]++;
            degree[v]++;
        }

        int leaf = 0;
        for (int i=2;i<=N;i++) {
            if (degree[i] == 1) {
                leaf++;
            }
        }

        System.out.println(W / leaf);

    }

}

File: 0_1.baekjoon/tree/kotlin/src/BOJ17037.kt
fun main() = with(System.`in`.bufferedReader()) {
    val (n, w) = readLine().split(" ").map { it.toInt() }
    val degree = Array(n + 1) { 0 }

    repeat(n - 1) {
        val (u, v) = readLine().split(" ").map { it.toInt() }
        degree[u]++
        degree[v]++
    }

    var leaf = 0
    for (i in 2..n) {
        if (degree[i] == 1) {
            leaf++
        }
    }

    print(w / leaf.toDouble())
}